drivers: clock_control: stm32h7 cannot choose system frequency higher than 400MHz

**Describe the bug**
It is not possible to configure a system clock higher than 400MHz on nucleo_h743zi board. It can also be the case for other H7 boards.
To achieve 400MHz+ frequencies, you need to use the PLL and set the Kconfig PLL parameters for the clock_control driver.

Moreover OpenOCD fails to complete flashing the device with status 1 and hello world is not shown in the console. In order to flash the device with other clock settings, it is necessary to spam the reset button while calling ```west flash``` to connect to the MCU under reset. See why this could happens below.

Fixing PR: #27213 

**To Reproduce**
Steps to reproduce the behavior:
1. Having a nucleo_h7 series board nucleo_h743zi preferably
2. Create a custom board named nucleo_h743zi_custom:
3. Modify the nucleo_h743zi_custom_defconfig file to change clock settings:
Comment existing settings and override them with:
```
# CSI PLL sysclk 480MHz
CONFIG_CLOCK_STM32_PLL_SRC_CSI=y
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000
CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=1
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=240
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2
CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```
or 
```
# HSI PLL sysclk 480MHz
CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000
CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=4
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=60
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2
CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```
or
```
# HSE PLL sysclk 480MHz
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=480000000
CONFIG_CLOCK_STM32_PLL_SRC_HSE=y
CONFIG_CLOCK_STM32_HSE_CLOCK=8000000
CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
CONFIG_CLOCK_STM32_HSE_BYPASS=y
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=240
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_R_DIVISOR=2
CONFIG_CLOCK_STM32_HPRE=2
CONFIG_CLOCK_STM32_D1CPRE=1
CONFIG_CLOCK_STM32_D2PPRE1=2
CONFIG_CLOCK_STM32_D2PPRE2=2
CONFIG_CLOCK_STM32_D1PPRE=2
CONFIG_CLOCK_STM32_D3PPRE=2
```
4. cd ~/zephyrproject/zephyr/samples/hello_world
4. west build -p always -b nucleo_h743zi_custom
3. west flash
4. See error

**Expected behavior**
Board should be flashed without any issues, Hello World should appear in the console and system clock should be set to 480MHz.

**Impact**
Prevents the user from setting whatever clock setting he wants. 
Could also break some h7 board like mentionned in #26980 .

**Logs and console output**
OpenOCD crash output:
```
[0/1] Flashing nucleo_h743zi_custom
-- west flash: using runner openocd
-- runners.openocd: Flashing file: /home/dev/eclipse-workspace/hello_world/build/zephyr/zephyr.hex
Open On-Chip Debugger 0.10.0+dev-01341-g580d06d9d-dirty (2020-05-16-15:41)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : clock speed 1800 kHz
Info : STLINK V2J29M18 (API v2) VID:PID 0483:374B
Info : Target voltage: 0.021961
Error: target voltage may be too low for reliable debugging
Info : stm32h7x.cpu0: hardware has 8 breakpoints, 4 watchpoints
Info : Listening on port 3333 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32h7x.cpu0      hla_target little stm32h7x.cpu       running

target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x08001248 msp: 0x240007d8
Info : Device: STM32H74x/75x
Info : flash size probed value 2048
Info : STM32H flash has dual banks. Bank (0) size is 1024kb, base address is 0x8000000
auto erase enabled
wrote 131072 bytes from file /home/dev/eclipse-workspace/hello_world/build/zephyr/zephyr.hex in 3.093589s (41.376 KiB/s)

target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x08001214 msp: 0x240007d8
verified 13852 bytes in 0.333224s (40.595 KiB/s)


FATAL ERROR: command exited with status 1: /home/dev/zephyr-sdk-0.11.3/sysroots/x86_64-pokysdk-linux/usr/bin/openocd -s /home/dev/zephyrproject/zephyr/boards/arm/nucleo_h743zi_custom/support -s /home/dev/zephyr-sdk-0.11.3/sysroots/x86_64-pokysdk-linux/usr/share/openocd/scripts -f /home/dev/zephyrproject/zephyr/boards/arm/nucleo_h743zi_custom/support/openocd.cfg -c init -c targets -c 'reset halt' -c 'flash write_image erase /home/dev/eclipse-workspace/hello_world/build/zephyr/zephyr.hex' -c 'reset halt' -c 'verify_image /home/dev/eclipse-workspace/hello_world/build/zephyr/zephyr.hex' -c 'reset run' -c shutdown
FAILED: zephyr/cmake/flash/CMakeFiles/flash 
cd /home/dev/eclipse-workspace/hello_world/build && /usr/bin/cmake -E env /home/dev/.local/bin/west flash --skip-rebuild
ninja: build stopped: subcommand failed.
```

**Environment (please complete the following information):**
 - OS: Linux Ubuntu 18.04 64bits VMWare Workstation 15.0
 - Toolchain: Zephyr SDK 0.11.3
 - Commit SHA : a4410367e5c81aa419e04532befc2a10e1b9e0c3

**Why this bug and Partial fix**
This bug occurs because some of the APB clocks can exceed the maximum allowed frequencies when PLL is being set as system source clock. In fact, the PLL dividers should be set to maximum prior to setting the PLL as system clock source and then correctly configured to prevent the frequencies to go out of the limits. 

About OpenOCD not being able to complete:
That must be due to the clock tree being stuck in an unspecified state preventing OpenOCD to complete all of its commands and exit with ```shutdown command invoked```. Because the flash writing operation succeeds in the first place the MCU must have booted the wrong clock configuration and frozen some of its clock buses  (APBx for instance) preventing OpenOCD to send commands like ```reset halt``` or ```reset run```.

I have a fix that allows the clock config from HSI/CSI/HSE + PLL to be flashed and used at 480MHz without any problems on nucleo_h743zi. 

**Additional context**
This problem appear for @Nukersson on his nucleo_h745zi_q board when trying to test #27188 .

--- a/drivers/clock_control/Kconfig.stm32h7
+++ b/drivers/clock_control/Kconfig.stm32h7
@@ -5,6 +5,17 @@
 
 if SOC_SERIES_STM32H7X
 
+# Oscillator clocks configuration options
+
+config CLOCK_STM32_HSI_DIVISOR
+	int "HSI Divisor"
+	depends on CLOCK_STM32_PLL_SRC_HSI || CLOCK_STM32_SYSCLK_SRC_HSI
+	default 1
+	range 1 8
+	help
+	  HSI Divisor to divide HSI base frequency value
+	  allowed values: 1, 2, 4, 8
+
 # Bus clocks configuration options
 
 config CLOCK_STM32_D1CPRE

--- a/drivers/clock_control/clock_stm32_ll_h7.c
+++ b/drivers/clock_control/clock_stm32_ll_h7.c
@@ -141,6 +141,80 @@ static uint32_t get_bus_clock(uint32_t clock, uint32_t prescaler)
 
 #if !defined(CONFIG_CPU_CORTEX_M4)
 
+static inline uint32_t get_pllsrc_frequency(void)
+{
+	switch (LL_RCC_PLL_GetSource()) {
+	case LL_RCC_PLLSOURCE_HSI:
+		return HSI_VALUE;
+	case LL_RCC_PLLSOURCE_CSI:
+		return CSI_VALUE;
+	case LL_RCC_PLLSOURCE_HSE:
+		return HSE_VALUE;
+	case LL_RCC_PLLSOURCE_NONE:
+	default:
+		return 0;
+	}
+}
+
+static uint32_t get_hclk_frequency(void)
+{
+	uint32_t sysclk = 0;
+	uint32_t hpre = 0;
+
+	/* Get the current system clock source */
+	switch (LL_RCC_GetSysClkSource()) {
+	case LL_RCC_SYS_CLKSOURCE_STATUS_HSI:
+		sysclk = HSI_VALUE;
+		break;
+	case LL_RCC_SYS_CLKSOURCE_STATUS_CSI:
+		sysclk = CSI_VALUE;
+		break;
+	case LL_RCC_SYS_CLKSOURCE_STATUS_HSE:
+		sysclk = HSE_VALUE;
+		break;
+	case LL_RCC_SYS_CLKSOURCE_STATUS_PLL1:
+		sysclk = PLLP_FREQ(get_pllsrc_frequency(),
+				   LL_RCC_PLL1_GetM(),
+				   LL_RCC_PLL1_GetN(),
+				   LL_RCC_PLL1_GetP());
+		break;
+	}
+	/* AHB/HCLK clock is sysclk/HPRE AHB prescaler*/
+	switch (LL_RCC_GetAHBPrescaler()) {
+	case LL_RCC_AHB_DIV_1:
+		hpre = 1;
+		break;
+	case LL_RCC_AHB_DIV_2:
+		hpre = 2;
+		break;
+	case LL_RCC_AHB_DIV_4:
+		hpre = 4;
+		break;
+	case LL_RCC_AHB_DIV_8:
+		hpre = 8;
+		break;
+	case LL_RCC_AHB_DIV_16:
+		hpre = 16;
+		break;
+	case LL_RCC_AHB_DIV_64:
+		hpre = 64;
+		break;
+	case LL_RCC_AHB_DIV_128:
+		hpre = 128;
+		break;
+	case LL_RCC_AHB_DIV_256:
+		hpre = 256;
+		break;
+	case LL_RCC_AHB_DIV_512:
+		hpre = 512;
+		break;
+	default:
+		hpre = 1;
+		break;
+	}
+	return get_bus_clock(sysclk, hpre);
+}
+
 static int32_t prepare_regulator_voltage_scale(void)
 {
 	/* Make sure to put the CPU in highest Voltage scale during clock configuration */
@@ -366,6 +440,8 @@ static int stm32_clock_control_init(struct device *dev)
 
 #if !defined(CONFIG_CPU_CORTEX_M4)
 	uint32_t pllsrc_clock = 0;
+	uint32_t old_hclk_freq = 0;
+	uint32_t new_hclk_freq = 0;
 
 #if defined(CONFIG_CLOCK_STM32_PLL_SRC_HSE) || \
 	defined(CONFIG_CLOCK_STM32_PLL_SRC_HSI) || \
@@ -388,6 +464,19 @@ static int stm32_clock_control_init(struct device *dev)
 
 	/* Configure Voltage scale to comply with the desired system frequency */
 	prepare_regulator_voltage_scale();
+
+	/* Current hclk value */
+	old_hclk_freq = get_hclk_frequency();
+	/* AHB is HCLK clock to configure */
+	new_hclk_freq = get_bus_clock(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC,
+				      CONFIG_CLOCK_STM32_HPRE);
+
+	/* Set flash latency */
+	/* AHB/AXI/HCLK clock is SYSCLK / HPRE */
+	/* If freq increases, set flash latency before any clock setting */
+	if (new_hclk_freq > old_hclk_freq) {
+		LL_SetFlashLatency(new_hclk_freq);
+	}
 
 	/* Configure PLL source */
 	/* Can be HSE , HSI 64Mhz/HSIDIV, CSI 4MHz*/
@@ -552,10 +641,11 @@ static int stm32_clock_control_init(struct device *dev)
 #endif /* CLOCK_STM32_SYSCLK_SRC */
 
 	/* Set FLASH latency */
-	/* AXI clock is SYSCLK / HPRE */
-	LL_SetFlashLatency(get_bus_clock(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC,
-									 CONFIG_CLOCK_STM32_HPRE));
-
+	/* AHB/AXI/HCLK clock is SYSCLK / HPRE */
+	/* If freq not increased, set flash latency after all clock setting */
+	if (new_hclk_freq <= old_hclk_freq) {
+		LL_SetFlashLatency(new_hclk_freq);
+	}
 
 	optimize_regulator_voltage_scale(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
 

--- a/drivers/clock_control/clock_stm32_ll_h7.c
+++ b/drivers/clock_control/clock_stm32_ll_h7.c
@@ -91,7 +91,7 @@ static int32_t get_vco_input_range(uint32_t pllsrc_clock, uint32_t divm)
 {
 	const uint32_t input_freq = pllsrc_clock/divm;
 
-	__ASSERT(input_freq < 1000000UL || input_freq > 16000000UL,
+	__ASSERT(input_freq >= 1000000UL && input_freq <= 16000000UL,
 			"PLL1 VCO frequency input range out of range");
 
 	if (1000000UL <= input_freq && input_freq <= 2000000UL) {
@@ -365,7 +365,7 @@ static int stm32_clock_control_init(struct device *dev)
 			pllsrc_clock,
 			CONFIG_CLOCK_STM32_PLL_M_DIVISOR);
 
-	__ASSERT(vco_input_range == -ERANGE, "PLL VCO input frequency out of range. Should be from 1 to 16 MHz");
+	__ASSERT(vco_input_range != -ERANGE, "PLL VCO input frequency out of range. Should be from 1 to 16 MHz");
 
 	vco_output_range = get_vco_output_range(vco_input_range);
 

--- a/drivers/clock_control/clock_stm32_ll_h7.c
+++ b/drivers/clock_control/clock_stm32_ll_h7.c
@@ -31,6 +31,97 @@
 #define z_apb4_prescaler(v) LL_RCC_APB4_DIV_ ## v
 #define apb4_prescaler(v) z_apb4_prescaler(v)
 
+/* Macro to check for clock feasability */
+/* It is Cortex M7's responsibility to setup clock tree */
+/* This check should only be performed for the M7 core code */
+#ifdef CONFIG_CPU_CORTEX_M7
+
+/* Define primary oscillator frequencies */
+/* Suppress the cast to uint32_t which */
+/* prevents from compare with #if > operator*/
+/* original defines in stm32h7xx_hal_conf.h*/
+#define HSI_FREQ 64000000UL	/* HSI_VALUE ((uint32_t)64000000) */
+/* HSE_VALUE overridden by the build system without C cast to uint*/
+/* Build system doesn't provide the UL type suffix for HSE_VALUE */
+/* Force HSE_FREQ to be cast to preprocessor UL to prevent overflow*/
+/* Only use these constants in preprocessor expressions */
+#define HSE_FREQ HSE_VALUE
+#define CSI_FREQ 4000000UL	/* CSI_VALUE ((uint32_t)4000000) */
+
+/* Choose PLL SRC */
+#ifdef CONFIG_CLOCK_STM32_PLL_SRC_HSI
+#define PLLSRC_FREQ  HSI_FREQ
+#elif defined(CONFIG_CLOCK_STM32_PLL_SRC_CSI)
+#define PLLSRC_FREQ  CSI_FREQ
+#elif defined(CONFIG_CLOCK_STM32_PLL_SRC_HSE)
+#define PLLSRC_FREQ  HSE_FREQ
+#else
+#define PLLSRC_FREQ 0
+#endif
+
+/* Given source clock and dividers, computed the output frequency of PLLP */
+#define PLLP_FREQ(pllsrc_freq, divm, divn, divp)	(((pllsrc_freq)*\
+							(divn))/((divm)*(divp)))
+
+/* PLL P output frequency value */
+#define PLLP_VALUE	PLLP_FREQ(\
+				PLLSRC_FREQ,\
+				CONFIG_CLOCK_STM32_PLL_M_DIVISOR,\
+				CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER,\
+				CONFIG_CLOCK_STM32_PLL_P_DIVISOR)
+
+/* SYSCLKSRC before the D1CPRE prescaler */
+#ifdef CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL
+#define SYSCLKSRC_FREQ	PLLP_VALUE
+#elif defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_HSI)
+#define SYSCLKSRC_FREQ	HSI_FREQ
+#elif defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_CSI)
+#define SYSCLKSRC_FREQ	CSI_FREQ
+#elif defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_HSE)
+#define SYSCLKSRC_FREQ	HSE_FREQ
+#endif
+
+/* ARM Sys CPU Clock before HPRE prescaler */
+#define SYSCLK_FREQ	((SYSCLKSRC_FREQ)/(CONFIG_CLOCK_STM32_D1CPRE))
+#define AHB_FREQ	((SYSCLK_FREQ)/(CONFIG_CLOCK_STM32_HPRE))
+#define APB1_FREQ	((AHB_FREQ)/(CONFIG_CLOCK_STM32_D2PPRE1))
+#define APB2_FREQ	((AHB_FREQ)/(CONFIG_CLOCK_STM32_D2PPRE2))
+#define APB3_FREQ	((AHB_FREQ)/(CONFIG_CLOCK_STM32_D1PPRE))
+#define APB4_FREQ	((AHB_FREQ)/(CONFIG_CLOCK_STM32_D3PPRE))
+
+/* Datasheet maximum frequency definitions */
+#define SYSCLK_FREQ_MAX	480000000UL
+#define AHB_FREQ_MAX	240000000UL
+#define APBx_FREQ_MAX	120000000UL
+
+#if SYSCLK_FREQ > SYSCLK_FREQ_MAX
+#error "SYSCLK frequency is too high, max is 480MHz"
+#endif
+#if AHB_FREQ > AHB_FREQ_MAX
+#error "AHB frequency is too high, max is 240MHz"
+#endif
+#if APB1_FREQ > APBx_FREQ_MAX
+#error "APB1 frequency is too high, max is 120MHz"
+#endif
+#if APB2_FREQ > APBx_FREQ_MAX
+#error "APB2 frequency is too high, max is 120MHz"
+#endif
+#if APB3_FREQ > APBx_FREQ_MAX
+#error "APB3 frequency is too high, max is 120MHz"
+#endif
+#if APB4_FREQ > APBx_FREQ_MAX
+#error "APB4 frequency is too high, max is 120MHz"
+#endif
+
+#if SYSCLK_FREQ != CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#error "Calculated CPU clock frequency (SYS clock) for M7 core doesn't match \
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC"
+#endif
+
+/* end of clock feasability check */
+#endif /* CONFIG_CPU_CORTEX_M7 */
+
+
 #if defined(CONFIG_CPU_CORTEX_M7)
 #if CONFIG_CLOCK_STM32_D1CPRE > 1
 /*

--- a/drivers/clock_control/clock_stm32_ll_h7.c
+++ b/drivers/clock_control/clock_stm32_ll_h7.c
@@ -13,6 +13,9 @@
 #include "stm32_hsem.h"
 
 /* Macros to fill up prescaler values */
+#define z_hsi_divider(v) LL_RCC_HSI_DIV ## v
+#define hsi_divider(v) z_hsi_divider(v)
+
 #define z_sysclk_prescaler(v) LL_RCC_SYSCLK_DIV_ ## v
 #define sysclk_prescaler(v) z_sysclk_prescaler(v)
 
@@ -50,7 +53,7 @@
 
 /* Choose PLL SRC */
 #ifdef CONFIG_CLOCK_STM32_PLL_SRC_HSI
-#define PLLSRC_FREQ  HSI_FREQ
+#define PLLSRC_FREQ  ((HSI_FREQ)/(CONFIG_CLOCK_STM32_HSI_DIVISOR))
 #elif defined(CONFIG_CLOCK_STM32_PLL_SRC_CSI)
 #define PLLSRC_FREQ  CSI_FREQ
 #elif defined(CONFIG_CLOCK_STM32_PLL_SRC_HSE)
@@ -74,7 +77,7 @@
 #ifdef CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL
 #define SYSCLKSRC_FREQ	PLLP_VALUE
 #elif defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_HSI)
-#define SYSCLKSRC_FREQ	HSI_FREQ
+#define SYSCLKSRC_FREQ	((HSI_FREQ)/(CONFIG_CLOCK_STM32_HSI_DIVISOR))
 #elif defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_CSI)
 #define SYSCLKSRC_FREQ	CSI_FREQ
 #elif defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_HSE)
@@ -160,11 +163,29 @@ static uint32_t get_hclk_frequency(void)
 {
 	uint32_t sysclk = 0;
 	uint32_t hpre = 0;
+	uint32_t hsidiv = 0;
+
+	/* Get the current HSI divider */
+	switch (LL_RCC_HSI_GetDivider()) {
+	case LL_RCC_HSI_DIV2:
+		hsidiv = 2;
+		break;
+	case LL_RCC_HSI_DIV4:
+		hsidiv = 4;
+		break;
+	case LL_RCC_HSI_DIV8:
+		hsidiv = 8;
+		break;
+	case LL_RCC_HSI_DIV1:
+	default:
+		hsidiv = 1;
+		break;
+	}
 
 	/* Get the current system clock source */
 	switch (LL_RCC_GetSysClkSource()) {
 	case LL_RCC_SYS_CLKSOURCE_STATUS_HSI:
-		sysclk = HSI_VALUE;
+		sysclk = HSI_VALUE/hsidiv;
 		break;
 	case LL_RCC_SYS_CLKSOURCE_STATUS_CSI:
 		sysclk = CSI_VALUE;
@@ -518,10 +539,8 @@ static int stm32_clock_control_init(struct device *dev)
 	while (LL_RCC_HSI_IsReady() != 1) {
 	}
 
-	/* Calibrate the HSI */
-	LL_RCC_HSI_SetCalibTrimming(32);
-	/* @TODO make HSI divider configurable */
-	LL_RCC_HSI_SetDivider(LL_RCC_HSI_DIV1);
+	/* HSI divider configuration */
+	LL_RCC_HSI_SetDivider(hsi_divider(CONFIG_CLOCK_STM32_HSI_DIVISOR));
 
 	/* Main PLL configuration and activation */
 	LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_HSI);
@@ -620,6 +639,9 @@ static int stm32_clock_control_init(struct device *dev)
 	LL_RCC_HSI_Enable();
 	while (LL_RCC_HSI_IsReady() != 1) {
 	}
+
+	/* HSI divider configuration */
+	LL_RCC_HSI_SetDivider(hsi_divider(CONFIG_CLOCK_STM32_HSI_DIVISOR));
 
 	/* Set sysclk source to HSI */
 	LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_HSI);

--- a/drivers/clock_control/clock_stm32_ll_h7.c
+++ b/drivers/clock_control/clock_stm32_ll_h7.c
@@ -398,13 +398,24 @@ static int stm32_clock_control_init(struct device *dev)
 	/* PLL will stay in reset state configuration */
 #endif /* CONFIG_CLOCK_STM32_PLL_SRC_* */
 
+
+	/* Preset the prescalers prior to chosing SYSCLK */
+	/* Prevents APB clock to go over limits */
+	/* Set buses (Sys,AHB, APB1, APB2 & APB4) prescalers */
+	LL_RCC_SetSysPrescaler(sysclk_prescaler(CONFIG_CLOCK_STM32_D1CPRE));
+	LL_RCC_SetAHBPrescaler(ahb_prescaler(CONFIG_CLOCK_STM32_HPRE));
+	LL_RCC_SetAPB1Prescaler(apb1_prescaler(CONFIG_CLOCK_STM32_D2PPRE1));
+	LL_RCC_SetAPB2Prescaler(apb2_prescaler(CONFIG_CLOCK_STM32_D2PPRE2));
+	LL_RCC_SetAPB3Prescaler(apb3_prescaler(CONFIG_CLOCK_STM32_D1PPRE));
+	LL_RCC_SetAPB4Prescaler(apb4_prescaler(CONFIG_CLOCK_STM32_D3PPRE));
+
+
 #if defined(CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL)
 
 	/* Enable PLL*/
 	LL_RCC_PLL1_Enable();
 	while (LL_RCC_PLL1_IsReady() != 1) {
 	}
-
 
 	/* Set PLL1 as System Clock Source */
 	LL_RCC_SetSysClkSource(LL_RCC_SYS_CLKSOURCE_PLL1);
@@ -448,14 +459,6 @@ static int stm32_clock_control_init(struct device *dev)
 	}
 
 #endif /* CLOCK_STM32_SYSCLK_SRC */
-
-	/* Set buses (Sys,AHB, APB1, APB2 & APB4) prescalers */
-	LL_RCC_SetSysPrescaler(sysclk_prescaler(CONFIG_CLOCK_STM32_D1CPRE));
-	LL_RCC_SetAHBPrescaler(ahb_prescaler(CONFIG_CLOCK_STM32_HPRE));
-	LL_RCC_SetAPB1Prescaler(apb1_prescaler(CONFIG_CLOCK_STM32_D2PPRE1));
-	LL_RCC_SetAPB2Prescaler(apb2_prescaler(CONFIG_CLOCK_STM32_D2PPRE2));
-	LL_RCC_SetAPB3Prescaler(apb3_prescaler(CONFIG_CLOCK_STM32_D1PPRE));
-	LL_RCC_SetAPB4Prescaler(apb4_prescaler(CONFIG_CLOCK_STM32_D3PPRE));
 
 	/* Set FLASH latency */
 	/* AXI clock is SYSCLK / HPRE */

